### PR TITLE
Refactor the voucher registry

### LINKED ISSUE #1742: refactor: Improvements to the Voucher Registry
This issue is to track improvements to the Voucher Registry to be tackled while the team is working all registries.

Parent Issue #1737.

-------

The follow changes should be made to the Voucher Registry:
 - [x] Make a "Menu" dropdown with the following options:
   1. Column Configuration
   2. Saving the Grid Configuration
   3. Resetting the Grid Configuration
   4. Download View as CSV
   5. Download View as PDF
 - [x] Implement the Grid State work on the grid (See #1738)
 - [x] Implement the latest FilterService, replacing the old `bhFiltersApplied` with the new `FilterService`
 - [x] Remove the `bhRendererDropdown`
 - [x] Implement downloading as CSV using the `ui-grid` exporter (See https://github.com/IMA-WorldHealth/bhima-2.X/pull/1554)

### PR DESCRIPTION
**Pull request by DedrickEnc**
This PR request is addressing the issue #1742 
Every mentioned on the issue has been addressed and the test was adapted
closes #1742 

Thank you for contributing!

Before submitting this pull request, please verify that you have:
 - [x] Run your code through ESLint.  [Check out our styleguide](https://github.com/IMA-WorldHealth/bhima-2.X/blob/master/docs/STYLEGUIDE.md).
 - [x] Run integration tests.
 - [x] Run end-to-end tests.
 - [x] Accurately described the changes your are making in this pull request.

For a more detailed checklist, [see the official review checklist](https://docs.google.com/document/d/1nupLVLRXgSZJQo_acLgrwvPnN8RukfSiwRhSToj81uU/pub) that this PR will be evaluated against.

Ensuring that the above checkboxes are completed will help speed the review process and help build a stronger application.  Thanks!

### CODE REVIEW

File: client/src/modules/vouchers/voucher-registry.html [hunk 1]
```diff
@@ -7,6 +7,58 @@
...
+
+            <li role="menuitem">
+              <a href data-method="clear-state" ng-click="VoucherCtrl.clearGridState()">
+                <i class="fa fa-close"></i> <span translate>FORM.BUTTONS.CLEAR_GRID_CONFIGURATION</span>
+              </a>
+            </li>
+
+            <li role="separator" class="divider"></li>
+            <li role="menuitem">
+              <a ng-href="/reports/finance/vouchers/?{{ VoucherCtrl.download('pdf') }}" download="{{ 'INVOICE_REGISTRY.TITLE' | translate }}">
```
Comment (sfount) [inline client/src/modules/vouchers/voucher-registry.html]: I think these keys are from the invoice registry - they should be changed for the voucher registry.

File: client/src/modules/vouchers/vouchers.service.js [hunk 2]
```diff
@@ -15,18 +15,78 @@ VoucherService.$inject = [
...
+  // service.formatFilterParameters = formatFilterParameters;
+  service.filters = voucherFilters;
+  service.cacheFilters = cacheFilters;
+  service.removeFilter = removeFilter;
+  service.loadCachedFilters = loadCachedFilters;
+  service.download = download;
+
+  voucherFilters.registerDefaultFilters([
+    { key : 'period', label : 'TABLE.COLUMNS.PERIOD', valueFilter : 'translate' },
+    { key : 'custom_period_start', label : 'PERIODS.START', valueFilter : 'date' },
```
Comment (sfount) [inline client/src/modules/vouchers/vouchers.service.js]: Custom period start and end should now have a comparison symbol to show that the dates are before and after the individual dates. 

They can be added like { comparitor : '<' }

(As of https://github.com/IMA-WorldHealth/bhima-2.X/pull/1774)

File: server/controllers/finance/cash.js [hunk 3]
```diff
@@ -154,11 +154,9 @@ function listPayment(options) {
   `;
 
   filters.period('period', 'date');
-  filters.dateFrom('custion_period_start', 'date');
+  filters.dateFrom('custom_period_start', 'date');
```
Comment (sfount) [inline server/controllers/finance/cash.js:157]: 👍  Good catch! FYI This was addressed in a very small PR here https://github.com/IMA-WorldHealth/bhima-2.X/pull/1774

**Review by sfount — COMMENTED** [review]
Code LGTM, I've made a couple of minor comments.

**Comment by DedrickEnc** [comment]
@sfount every test have passed

**Comment by sfount** [comment]
Everything looks good here - it looks like the filters are not cached directly when the user updates the search - either that or they are not loaded when the page is loaded. 

If that can be quickly addressed this can be directly merged.

**Comment by DedrickEnc** [comment]
@sfount
The bug was fixed

File: client/src/modules/vouchers/modals/search.modal.js [hunk 4]
```diff
@@ -13,59 +14,76 @@ VoucherRegistrySearchModalController.$inject = [
...
   var vm = this;
+  var changes = new Store({identifier : 'key'});
+  vm.filters = filters;
+  vm.searchQueries = {};
+  vm.defaultQueries = {};
 
-  // set controller data
-  vm.params = angular.copy(filters || {});
+  var searchQueryOptions = [
+    'reference', 'description', 'user_id',
```
Comment (sfount) [inline client/src/modules/vouchers/modals/search.modal.js]: Adding `trans_id` here will allow the transaction ID to be loaded when the user opens the search modal for a second time.

**Review by sfount — APPROVED** [review]
Changes LGTM :+1:. Searching works as expected, great work!

**Pull request merged**